src/geom/density.jl

```diff
@@ -0,0 +1,191 @@
+struct DensityGeometry <: Gadfly.GeometryElement
+    stat::Gadfly.StatisticElement
+    order::Int
+    tag::Symbol
+end
+
+function DensityGeometry(; n=256,
+                           bandwidth=-Inf,
+                           adjust=1.0,
+                           kernel=Normal,
+                           trim=false,
+                           scale=:area,
+                           position=:dodge,
+                           orientation=:horizontal,
+                           order=1,
+                           tag=empty_tag)
+    stat = Gadfly.Stat.DensityStatistic(n, bandwidth, adjust, kernel, trim,
+                                        scale, position, orientation, false)
+    DensityGeometry(stat, order, tag)
+end
+
+DensityGeometry(stat; order=1, tag=empty_tag) = DensityGeometry(stat, order, tag)
+
+"""
+   Geom.density(; bandwidth, adjust, kernel, trim, scale, position, orientation, order)
+
+Draws a kernel density estimate. This is a cousin of [`Geom.histogram`](@ref)
+that is especially useful when the datapoints originate from a underlying smooth
+distribution. Unlike histograms, density estimates do not suffer from edge
+effects from incorrect bin choices. Some caveats do apply:
+
+1) Plot components do not necessarily correspond to the raw datapoints, but
+   instead to the kernel density estimation of the underlying distribution
+2) Density estimation improves as a function of the number of data points and
+   can be misleadingly smooth when the number of datapoints is small.
+3) Results can be sensitive to the choise of `kernel` and `bandwidth`
+
+For horizontal histograms (default), `Geom.density` draws the kernel density
+estimate of `x` optionally grouped by `color`. If the `orientation=:vertical`
+flag is passed to the function, then densities will be computed along `y`. The
+estimates are normalized by default to have areas equal to 1, but this can
+changed by passing `scale=:count` to scale by the raw number of datapoints or
+`scale=:peak` to scale by the max height of the estimate. Additionally, multiple
+densities can be stacked using the `position=:stack` flag or the conditional
+density estimate can be drawn using `position=:fill`. See
+[`Stat.DensityStatistic`](@ref Gadfly.Stat.DensityStatistic) for details on
+optional parameters that can control the `bandwidth`, `kernel`, etc used.
+
+External links
+
+* [Kernel Density Estimation on Wikipedia](https://en.wikipedia.org/wiki/Kernel_density_estimation)
+"""
+const density = DensityGeometry
+
+element_aesthetics(::DensityGeometry) = Symbol[]
```

Comment:
element_aesthetics should contain `:x`, `:y`, and `:color`, no?

Reply:
If I don't leave this blank, they are filled with autogenerated values so it's impossible to give useful error messages using `Gadfly.assert_aesthetics_defined`. I wasn't sure how to get around this so I leave this blank and figure out errors later: https://github.com/GiovineItalia/Gadfly.jl/pull/1157/files#diff-9ec506bf78232ae17d082c22c2e66449R616